File: graph_aeron_flatbuffers/cpp/src/flatbufferutils/createBook.cpp
#include "createBook.h"

#include "../../generated/book_generated.h"

namespace Graph {
namespace FlatBufferUtils {

std::vector<BufferT> AsBuffer() {
    flatbuffers::FlatBufferBuilder builder;

    auto instId = Graph::CreateInstrumentIdDirect(builder, Graph::Exchange::Exchange_Deribit, "InstId1");
    Graph::Timestamp timestamp(100);
    Graph::BookUpdateBuilder bub(builder);
    bub.add_timestamp(&timestamp); 
    bub.add_id(instId);
    auto offset = bub.Finish();
    Graph::FinishBookUpdateBuffer(builder, offset);

    auto bufferPointer = builder.GetBufferPointer();

    std::vector<BufferT> buf;
    buf.insert(buf.end(), bufferPointer, bufferPointer + builder.GetSize());
    return buf;
}

}
}

File: graph_aeron_flatbuffers/rust/src/flatbufferutils/create_book.rs
extern crate flatbuffers;

#[path = "../../generated/common_generated.rs"]
mod common_generated;

#[path = "../../generated/book_generated.rs"]
mod book_generated;

// pub mod flat_buffer_utils {
    use crate::{ InstrumentId, InstrumentIdArgs, Exchange, Timestamp };
    use crate::{ book_generated::graph::BookUpdateBuilder };

    pub fn as_buffer() -> Vec<u8> {
        let mut builder = flatbuffers::FlatBufferBuilder::with_capacity(1024);
        let inst_id_str = builder.create_string("InstId1");
        
        let inst_id_args = InstrumentIdArgs {
            exchange: Exchange::Deribit,
            id: Some(inst_id_str)
        };
    
        let inst_id = InstrumentId::create(&mut builder, &inst_id_args);
        let timestamp = Timestamp::new(100);
    
        let mut bub = BookUpdateBuilder::new(&mut builder);
        bub.add_timestamp(&timestamp);
        bub.add_id(inst_id);
        let offset = bub.finish();
        let pointer = builder.finish(offset, None);
        let fin_buf = builder.finished_data();
        fin_buf.to_vec()
    }
// }